ext/har/logger.go

```diff
@@ -0,0 +1,157 @@
+func WithExportThreshold(threshold int) LoggerOption {
+    return func(l *Logger) {
+        l.exportThreshold = threshold
+    }
+}
+
+// NewLogger creates a new HAR logger instance
+func NewLogger(exportFunc ExportFunc, opts ...LoggerOption) *Logger {
+    l := &Logger{
+        entries:        make([]Entry, 0),
+        captureContent: true,
+        exportFunc:     exportFunc,
+        exportThreshold: 100,    // Default threshold
+        exportInterval: 0,       // Default no interval
+        stopCh:         make(chan struct{}),
+        dataCh:         make(chan Entry, 1000), 
+    }
+    
+    // Apply options
+    for _, opt := range opts {
+        opt(l)
+    }
+    
+    go l.exportLoop()
+    return l
+}
+
+// OnRequest handles incoming HTTP requests
+func (l *Logger) OnRequest(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
+    if ctx != nil {
+        ctx.UserData = time.Now()
+    }
+    return req, nil
+}
+
+// OnResponse handles HTTP responses
+func (l *Logger) OnResponse(resp *http.Response, ctx *goproxy.ProxyCtx) *http.Response {
+    if resp == nil || ctx == nil || ctx.Req == nil || ctx.UserData == nil {
+        return resp
+    }
+    startTime, ok := ctx.UserData.(time.Time)
+    if !ok {
+        return resp
+    }
+    
+    entry := Entry{
+        StartedDateTime: startTime,
+        Time:           time.Since(startTime).Milliseconds(),
+        Request:        ParseRequest(ctx, l.captureContent),
+        Response:       ParseResponse(ctx, l.captureContent),
+        Timings: Timings{
+            Send:    0,
+            Wait:    time.Since(startTime).Milliseconds(),
+            Receive: 0,
+        },
+    }
+    entry.fillIPAddress(ctx.Req)
+    
+    select {
+    case l.dataCh <- entry:
+    default:
+        // Log or handle case where channel is full, just in case 
+        ctx.Proxy.Logger.Printf("Warning: HAR logger channel is full, dropping entry")
+    } 
+    
+    return resp
+}
+
+func (l *Logger) processEntry(entry Entry) {
+    l.entries = append(l.entries, entry)
+    if l.exportThreshold > 0 && len(l.entries) >= l.exportThreshold {
+        l.exportFunc(l.entries)
+        l.entries = make([]Entry, 0)
+    }
+}
+
+func (l *Logger) exportIfNeeded() {
+    if len(l.entries) > 0 {
+        l.exportFunc(l.entries)
+        l.entries = make([]Entry, 0)
+    }
+}
+
+func (l *Logger) drainChannel() {
+    for {
+        select {
+        case entry := <-l.dataCh:
+            l.processEntry(entry)
+        default:
+            return
+        }
+    }
+}
+
+func (l *Logger) exportLoop() {
+    var tickerC <-chan time.Time
+    if l.exportInterval > 0 {
+        ticker := time.NewTicker(l.exportInterval)
+        defer ticker.Stop()
+        tickerC = ticker.C
+    }
+    
+    for {
+        select {
+        case <-l.stopCh:
+            l.drainChannel()
+            l.exportIfNeeded()
+            return
```

Comment:
Move this to dataCh when ok value is false.
In that case, the data channel is closed and you just pass the existing slice to the entries handler. No need to use a custom drainChannel or something like this.